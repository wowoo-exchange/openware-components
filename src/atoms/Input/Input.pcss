.cr-input {
  display: inline-block;
  background-color: inherit;

  input {
    border: none;
    outline: none;
    text-align: left;
    background-color: transparent;
    color: var(--base-1);
  }

  &__number {
    input {
      &:focus {
        margin: 0;
      }
    }
  }
}
