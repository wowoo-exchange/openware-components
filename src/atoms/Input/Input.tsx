import classnames from 'classnames';
import * as React from 'react';
import { OnChangeEvent } from '../../types/index';

interface InputProps {
    /**
     * Function for monitoring changes of input value
     * @default ''
     */
    onChangeValue: (text: string) => void;
    /**
     * Function for handling input focus
     */
    onFocus?: () => void;
    /**
     * Function for handling input blur
     */
    onBlur?: () => void;
    /**
     * The text string to use for the type of the input component.
     */
    value: string | number;
    /**
     * Additional class name for styling. By default element receives `cr-input` class
     * @default empty
     */
    className?: string;
    /**
     * The string to use for as a placeholder for the input
     * @default empty
     */
    placeholder?: string;
    /**
     * Sets type of the input
     * @default 'text'
     */
    type?: string;
    /**
     * Input field name
     */
    name?: string;
    /**
     * Set autofocus on field
     */
    autoFocus?: boolean;
    /**
     * Function for handling 'Enter' key
     */
    onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
}

/**
 * Cryptobase Input that overrides default input
 */
class Input extends React.PureComponent<InputProps> {

    private inputElem = React.createRef<HTMLInputElement>();
    public render() {
        const {
            placeholder,
            type,
            className,
            value,
            name,
            autoFocus,
            onKeyPress,
        } = this.props;

        const cx = classnames('cr-input', {
            'cr-input__number': type === 'number',
        }, className);

        const autoComplete = type === 'password' ? 'current-password' : '';
        return (
            <div className={cx}>
                <input
                    min="0"
                    name={name}
                    autoComplete={autoComplete}
                    type={type}
                    ref={this.inputElem}
                    placeholder={placeholder}
                    value={value}
                    onChange={this.handleChange}
                    onFocus={this.props.onFocus}
                    onBlur={this.props.onBlur}
                    autoFocus={autoFocus}
                    onKeyPress={onKeyPress}
                />
            </div>
        );
    }

    private handleChange = (e: OnChangeEvent) => {
        this.props.onChangeValue(e.target.value);
    };
}

export {
    Input,
    InputProps,
};
