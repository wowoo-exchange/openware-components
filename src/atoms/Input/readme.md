Example of usage:

```js
  const handleChange = () => undefined;
  const onKeyPress = (event) => {
      if (event.key === "Enter") {
        console.log("Enter");
      }
  }; 

  <Input
    type="text"
    placeholder="Placeholder"
    min="0"
    onChangeValue={handleChange}
    onKeyPress={onKeyPress}
    />
```