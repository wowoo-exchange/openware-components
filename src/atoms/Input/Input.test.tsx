import { shallow } from 'enzyme';
import * as React from 'react';
import { SinonSpy, spy } from 'sinon';
import { Input, InputProps } from './Input';


const defaultProps: InputProps = {
    onChangeValue: () => undefined,
    type: 'text',
    value: 'value',
};

const setup = (props: Partial<InputProps> = {}) =>
    shallow(<Input {...{ ...defaultProps, ...props }} />);

describe('Input', () => {
    it('should render', () => {
        const wrapper = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        const wrapper = setup();
        expect(wrapper.hasClass('cr-input')).toBeTruthy();
    });

    it('should pass along supplied className', () => {
        const className = 'some-class';
        const wrapper = setup({ className });
        expect(wrapper.hasClass(className)).toBeTruthy();
    });

    it('should have correct className when type is number', () => {
        defaultProps.type = 'number';
        const wrapper = setup();
        expect(wrapper.hasClass('cr-input__number')).toBeTruthy();
    });

    it('should have correct placeholder', () => {
        const placeholder = 'I am Placeholder';
        const wrapper = setup({ placeholder });
        const { children } = wrapper.props();
        expect(children.props.placeholder).toEqual(placeholder);
    });

    it('should handle onChangeValue callback', () => {
        const onChangeValue = spy();
        const wrapper = setup({ onChangeValue });
        wrapper.find('input').simulate('change', {target: {value: 'My new value'}});
        expect((onChangeValue as SinonSpy).calledOnceWith('My new value')).toBeTruthy();
    });

    it('should handle input "-"', () => {
        const wrapper = setup();
        const wrapperBefore = wrapper;
        const event = new KeyboardEvent('keyPress', {key : '-'});
        wrapper.find('input').simulate('keyPress', event);
        expect(wrapperBefore).toEqual(wrapper);
    });

    it('should handle input "."', () => {
        const wrapper = setup({ value: '1.2' });
        const wrapperBefore = wrapper;
        const event = new KeyboardEvent('keyPress', {key : '.'});
        wrapper.find('input').simulate('keyPress', event);
        expect(wrapperBefore).toEqual(wrapper);
    });

    it('should autoComplete input', () => {
        const wrapper = setup();
        expect(wrapper.props().children.props.autoComplete).toEqual('');
        wrapper.setProps({ type: 'password' });
        expect(wrapper.props().children.props.autoComplete).toEqual('current-password');
    });

    it('should handle onKeyPress event', () => {
        const onKeyPress = spy();
        const wrapper = setup({ onKeyPress });
        const wrapperBefore = wrapper;
        const event = new KeyboardEvent('keyPress', { key: 'Enter' });
        wrapper.find('input').simulate('keyPress', event);
        expect(wrapperBefore).toEqual(wrapper);
    });
});
