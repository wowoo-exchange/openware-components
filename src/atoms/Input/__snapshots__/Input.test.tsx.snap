// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render 1`] = `
<div
  className="cr-input"
>
  <input
    autoComplete=""
    min="0"
    onChange={[Function]}
    type="text"
    value="value"
  />
</div>
`;
